File: cpp16/ex4/find_if.cpp
#include <iostream>
#include <Vector>

template<typename T>
class Matcher{
	public:
		bool match(T elem){
		if(true){
			std::cout << "matched";
			return true;
		}
	}
};
template<>
class Matcher<std::string>{
	public:
		bool match(std::string elem){
		if(elem.compare("Hello") == 0){
			std::cout << "matched";
			return true;
		}
	}
};

template<typename T>
	bool findif(typename std::vector<T>::iterator begin,typename std::vector<T>::iterator end, Matcher<T> m){ 
		while(begin != end){
			if(m.match(*begin)){return true;}
			*begin++;
		}
	}
	
int main(){
	std::vector<std::string> vector(1000000,"Hello");
	Matcher<std::string> matcher;
	std::vector<std::string>::iterator begin = vector.begin();
	std::vector<std::string>::iterator end = vector.end();

	findif<std::string>(begin,end,matcher);
	
	
	return 0;
}

File: cpp16/ex4/Find_if.java
package find_if;
import java.util.Iterator;
import java.util.Vector;

public class Find_if implements Matcher<String> { //Apparently you can't implement a generic interface
	public static void main(String[] args){		// if class is not generic
			Vector<String> vec = new Vector();
			for(int i=0;i<1000000;i++){
				vec.add("Hello");
			}
			Iterator itr = vec.iterator();
			
	}
	public void find_if(Iterator it,Matcher match){
		while(it.hasNext()){
			 if(match(it.next().toString())){//Something wrong here
				 System.out.println("Matched");				 
			 }
		}
			
	}

	@Override
	public boolean match(String elem){
		if(elem.equals("Hello")){
		return true;
		}
		return false;
	}
}
